llvm/lib/CAS/OnDiskCASLogger.cpp

```diff
@@ -0,0 +1,223 @@
+//===- OnDiskCASLogger.cpp ------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/CAS/OnDiskCASLogger.h"
+
+#include "llvm/ADT/StringSwitch.h"
+#include "llvm/ADT/Twine.h"
+#include "llvm/Config/llvm-config.h"
+#include "llvm/Support/Error.h"
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/raw_ostream.h"
+#if LLVM_ON_UNIX
+#include <sys/time.h>
+#include <unistd.h>
+#endif
+
+using namespace llvm;
+using namespace llvm::cas;
+using namespace llvm::cas::ondisk;
+
+// The version number in this log should be bumped if the log format is changed
+// in an incompatible way. It is currently a human-readable text file, so in
+// practice this would be if the log changed to binary or other machine-
+// readable format.
+static constexpr StringLiteral Filename = "/v1.log";
+
+OnDiskCASLogger::OnDiskCASLogger(raw_fd_ostream &OS, bool LogAllocations)
+    : OS(OS), LogAllocations(LogAllocations) {}
+
+OnDiskCASLogger::~OnDiskCASLogger() {
+  OS.flush();
+  delete &OS;
+}
+
+static bool isDisabledEnv(StringRef V) {
+  return StringSwitch<bool>(V)
+      .Case("0", true)
+      .CaseLower("no", true)
+      .CaseLower("false", true)
+      .Default(false);
+}
+
+Expected<std::unique_ptr<OnDiskCASLogger>>
+OnDiskCASLogger::openIfEnabled(const Twine &Path) {
+  const char *V = getenv("LLVM_CAS_LOG");
+  if (V && !isDisabledEnv(V)) {
+    int LogLevel = -1;
+    StringRef(V).getAsInteger(10, LogLevel);
+    return OnDiskCASLogger::open(Path, /*LogAllocations=*/LogLevel > 1 ? true
+                                                                       : false);
+  }
+  return nullptr;
+}
+Expected<std::unique_ptr<OnDiskCASLogger>>
+OnDiskCASLogger::open(const Twine &Path, bool LogAllocations) {
+  std::error_code EC;
+  SmallString<128> FullPath;
+  (Path + Filename).toVector(FullPath);
```

Comment:
why not using `sys::path::append()` and make Filename not starting with `/`. That will be more portable across systems.